sql/src/main/java/org/apache/druid/sql/http/SqlResource.java

```diff
@@ -205,22 +205,20 @@ public Response doPost(
       throw (ForbiddenException) serverConfig.getErrorResponseTransformStrategy()
                                              .transformIfNeeded(e); // let ForbiddenExceptionMapper handle this
     }
+    catch (RelOptPlanner.CannotPlanException e) {
+      endLifecycle(sqlQueryId, lifecycle, e, remoteAddr, -1);
+      SqlPlanningException spe = new SqlPlanningException(SqlPlanningException.PlanningError.UNSUPPORTED_SQL_ERROR,
+          e.getMessage());
+      return buildNonOkResponse(BadQueryException.STATUS_CODE, spe, sqlQueryId);
```

Comment:
i had to think a bit on this because I was a bit torn. One the one hand, it's not necessarily the users fault that we don't support some SQL syntax they might have tried so a server error seems fair in that sense, but on the other hand being a client error might inspire the user to try to rewrite their query where a server error might not, so I've come around to being :+1: on this

Reply:
I think it is weird currently that the HTTP error code that the user gets depends on when can we detect that the query is bad or unsupported. I wanted the behaviour to be consistent irrespective of the stage, the query gets rejected. 
If there is an unknown failure (NPE, ISE) during the detection path, that should be an internal server error but if we did detect it correctly that query is unplannable then we know we can't support the said query syntax. this is why 400 is returned only in the case of CannotPlanException. 

Like a user trying to do max aggregation on a string or a user trying to use multiple exact distinct counts will get 500 right now even though it is documented that these features are not supported. these are indeed bad queries in that sense. Also, what matters more than the error code, is the info we are giving alongside in the response since that is more useful for the end-user.